### PR TITLE
Update the Plugin Manager Updates View with more Info about incompatible dependencies and change CSS classes

### PR DESCRIPTION
**Pull request by andipabst**
This is a hacktoberfest contribution.

### Proposed changelog entries

* Add information about incompatible dependencies to the updates view in the plugin manager
* Change the classes for the warnings in the updates view to use 'alert-danger'

Before:
Bildschirmfoto am 2019-10-16 um 16 26 26

After: Bildschirmfoto am 2019-10-16 um 16 22 29

### CODE REVIEW

File: core/src/main/resources/hudson/PluginManager/_table.css
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/_table.css]: add line break at the end

```suggestion
}

```

File: war/src/main/webapp/css/style.css
Comment (oleg-nenashev) [inline war/src/main/webapp/css/style.css]: Have you checked they are not used anywhere else? It might make sense to just keep them as deprecated. 
CC @varyvol @batmat @raul-arabaolaza since some proprietary components might depend on that, not checked

File: core/src/main/resources/hudson/PluginManager/table.properties
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/table.properties]: A new entry might be better so that localization tools notice it

**Review by oleg-nenashev — COMMENTED** [review]
Thanks! Just few minor comments

File: war/src/main/webapp/css/style.css
Comment (andipabst) [reply war/src/main/webapp/css/style.css]: Yes, I searched through the repository and could not find them used anywhere else. 

How would I mark them as deprecated? Just with a CSS comment or in any special way?

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (zbynek) [inline core/src/main/java/hudson/model/UpdateSite.java]:
```suggestion
                    if (!p.isCompatibleWithInstalledVersion() || !p.isNeededDependenciesCompatibleWithInstalledVersion(cache)) {
```

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (zbynek) [inline core/src/main/java/hudson/model/UpdateSite.java]: If we now cache the whole list of dependencies, we don't need to also cache the `bool` whether it's empty.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (zbynek) [inline core/src/main/java/hudson/model/UpdateCenter.java]: This can be simplified using `incompatiblePluginMap.computeIfAbsent`.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (zbynek) [inline core/src/main/java/hudson/model/UpdateCenter.java]: If we would index this map with `Plugin` instead of plugin name, we could avoid lookup in `pluginMap`.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (zbynek) [inline core/src/main/java/hudson/model/UpdateSite.java]: Why not always use the version that requires a cache?

**Review by zbynek — COMMENTED** [review]
The new look is nice!

File: war/src/main/webapp/css/style.css
Comment (oleg-nenashev) [reply war/src/main/webapp/css/style.css]: Let's wait for confirmation from CCed users. If they confirm there is no external dependency, IMHO we are good. It is highly unlikely that there are other companies which might be affected, and having a single alarm type is net positive IMHO

Comment (oleg-nenashev) [reply war/src/main/webapp/css/style.css]: Regarding marking as deprecated, yes we can just put a CSS comment. It is better to add additional comment for removing it, e.g. `TODO(andipabst): Not used after https://github.com/jenkinsci/jenkins/pull/4299, remove once there is no dependencies`

Comment (varyvol) [reply war/src/main/webapp/css/style.css]: I'm afraid we are using this internally.

Comment (oleg-nenashev) [reply war/src/main/webapp/css/style.css]: @andipabst Then we just revert it and "deprecate" in Comments, I'd guess

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (andipabst) [reply core/src/main/java/hudson/model/UpdateSite.java]: Tbh I just implemented it like in e.g. `isNeededDependenciesCompatibleWithInstalledVersion()` and `isNeededDependenciesForNewerJenkins()`, because I thought the versions of the methods without parameters would be necessary. But I can of course remove this method, if they are not needed?

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (andipabst) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Oh thanks, I didn't know that method!

File: core/src/main/resources/hudson/PluginManager/_table.css
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/_table.css]: hm? Same as other classes?

Comment (andipabst) [reply core/src/main/resources/hudson/PluginManager/_table.css]: Oh, I thought they would be specific to the table and also just be usable there, since they are not in a global CSS file but in the PluginManager directory. Should this be changed as well?

Comment (oleg-nenashev) [reply core/src/main/resources/hudson/PluginManager/_table.css]: I think you just know CSS better than me :) If it is an internal CSS, my comment is definitely invalid. Thanks for clarification!

**Review by oleg-nenashev — APPROVED** [review]
Looks good from the behavior/compatibility standpoint. I am not that concerned about performance for this page, but it definitely needs discussion with @zbynek to be finished before merging

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (zbynek) [reply core/src/main/java/hudson/model/UpdateSite.java]: If you add a new method, you create public API visible to plugins to be maintained forever. So if the goal is not to use it from plugins, I'd suggest to remove it.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (zbynek) [reply core/src/main/java/hudson/model/UpdateSite.java]: The code is a modified copy of the methods introduced in https://github.com/jenkinsci/jenkins/pull/3428 , co not passing the cache argument is not a new isssue. However if we pass it, we can increase the hit ratio from 15% to 72%   CC @jglick

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (andipabst) [reply core/src/main/java/hudson/model/UpdateSite.java]: Okay, thank you for your explanation!

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (andipabst) [reply core/src/main/java/hudson/model/UpdateSite.java]: Okay, I will rewrite the boolean version to just check whether the list is empty.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (andipabst) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Could you please explain this further to me?
If `incompatiblePluginMap` would be indexed with `Plugin`, how would I iterate through the `pluginMap` later and add the incompatible plugins there?

**Review by zbynek — APPROVED** [review]
Thanks for checking the suggestions!

**Review by fcojfernandez — APPROVED** [review]

**Review by alecharp — APPROVED** [review]
I find the new UI less aggressive and still more noticeable than the old one. Thanks!

**Comment by fcojfernandez** [comment]
thanks @andipabst ! I've added the ready for merge label.

**Comment by oleg-nenashev** [comment]
Thanks a lot for your contributions @andipabst ! It is much appreciated, and we invite you to keep contributing to the project. 👍

**Comment by andipabst** [comment]
Thank you all for your help in this pull request. I would very much like to keep contributing!

**Pull request merged**
